src/ostrowski/rationals/basic.lean

```diff
@@ -0,0 +1,226 @@
+  (same_on_neg_one: φ₁ (-1) = φ₂ (-1)) (same_on_nat: ∀ n: ℕ, φ₁ n = φ₂ n): (φ₁: ℚ → ℝ) = φ₂ :=
+begin
+  ext,
+  suffices : ∀ z : ℤ, φ₁ z = φ₂ z,
+  begin
+    induction x,
+    rw rat_mk_eq_div,
+    repeat { rw div_eq_mul_inv, },
+    have x_denom_ne_zero: (x_denom: ℚ) ≠ 0,
+    {
+      symmetry,
+      norm_cast,
+      exact (ne_of_lt x_pos),
+    },
+    rw monoid_with_zero_hom.map_mul φ₁,
+    rw Γ₀_map_inv φ₁ x_denom x_denom_ne_zero,
+    rw monoid_with_zero_hom.map_mul φ₂,
+    rw Γ₀_map_inv φ₂ x_denom x_denom_ne_zero,
+    rw this x_num,
+    have := this (↑ x_denom),
+    norm_cast at this,
+    rw this,
+  end,
+  intro x,
+  suffices : ∀ z : ℕ, φ₁ z = φ₂ z,
+  begin
+    induction x,
+    simp [this],
+    push_cast,
+    conv {
+      congr,
+      rw neg_eq_neg_one_mul,
+      skip,
+      rw neg_eq_neg_one_mul,
+      skip,
+    },
+    rw monoid_with_zero_hom.map_mul φ₁,
+    rw monoid_with_zero_hom.map_mul φ₂,
+    rw same_on_neg_one,
+    simp,
+    norm_cast,
+    exact this _,
+  end,
+  intro x,
+  cases x,
+  simp,
+  have := same_on_nat (nat.succ x_1),
+  simp [-add_comm] at this,
+  simp [-add_comm],
+  rw this,
+end
+
+lemma abs_val_equiv_of_equiv_on_primes (abv abv': ℚ → ℝ)
+  [habv: is_absolute_value abv] [habv': is_absolute_value abv']
+  (α: ℝ) (α_ne_zero: α ≠ 0)
+  (h: ∀ p: ℕ, prime p → (abv p) ^ α = abv' p): (λ r: ℚ, (abv r) ^ α) = abv' :=
+begin
+  have: ∀ n: ℕ, (abv n) ^ α = abv' n,
+  {
+    have inductive_step: ∀ q a: ℕ, prime q → (abv a) ^ α = abv' a
+      → (abv (q * a: ℕ)) ^ α = abv' ((q * a): ℕ),
+    {
+      intros q a q_prime a_norms_eq,
+      calc (abv (q * a: ℕ)) ^ α
+        = abv (q*a: ℕ) ^ α
+          : by { refl }
+        ... = ((abv q) * (abv a)) ^ α
+          : by { push_cast, rw abv_mul abv, }
+        ... = (abv q) ^ α * (abv a) ^ α
+          : by { rw real.mul_rpow (abv_nonneg abv q) (abv_nonneg abv a), }
+        ... = (abv' q) * (abv' a)
+          : by { rw [h q q_prime, a_norms_eq], }
+        ... = abv' (q * a)
+          : by { rw ← abv_mul abv', }
+        ... = abv' ((q * a): ℕ)
+          : by { norm_cast, },
+    },
+    apply induction_on_primes,
+    {
+      norm_cast,
+      rw [abv_zero abv, abv_zero abv'],
+      rw real.zero_rpow,
+      exact α_ne_zero,
+    },
+    {
+      norm_cast,
+      rw [abv_one abv, abv_one abv'],
+      rw real.one_rpow,
+    },
+    exact inductive_step,
+  },
+  
+  set equiv_hom := hom_of_equiv abv α α_ne_zero with equiv_hom_def,
+  set abv'_hom := abv_hom abv' with abv'_hom_def,
+  have abv_eq_hom_fn: abv' = abv'_hom := rfl,
+  have equiv_abs_eq_hom_fn: (λ r: ℚ, abv r ^ α) = equiv_hom := rfl,
+  have same_on_neg_one: equiv_hom (-1) = abv'_hom (-1),
+  {
+    rw [← abv_eq_hom_fn, ← equiv_abs_eq_hom_fn],
+    simp,
+    rw [abv_neg abv, abv_neg abv'],
+    rw [abv_one abv, abv_one abv'],
+    rw real.one_rpow,
+  },
+  apply mul_mor_eq_of_eq_on_pnat equiv_hom abv'_hom same_on_neg_one,
+
```

Comment:
kill this line

Reply:
I like this line :'(
Actually, it has a reason to exist : the final `exact` is disconnected from what precedes (which is a block that should actually be rewritten by properly using typeclasses).